lib/nx_deflate.c

```diff
@@ -933,10 +971,38 @@ static int  nx_compress_block_update_offsets(nx_streamp s, int fc)
 	   update the input pointers
 	*/
 
-	ASSERT(s->spbc >= s->used_in);
-	int from_next_in = s->spbc - s->used_in;
-	s->used_in = 0;
-	s->cur_in = 0;
+	from_used_in = NX_MIN(s->used_in, s->spbc);
+	from_next_in = s->spbc > s->used_in ? s->spbc - s->used_in : 0;
+
+	s->cur_in += from_used_in;
+	s->used_in -= from_used_in;
+
+	/* Make sure fifo_in has been allocated */
+	fifo_in_alloc_check(s);
+
+	/* We need to copy data from next_in to fifo_in to maintain the history. We
+	   also need to limit the amount of data in fifo_in to len_in/2, so we may
+	   need to throw away old data if we reach that limit. */
+	if (s->cur_in + from_next_in > s->len_in/2) {
+
+		/* Throwing away old data. We just want to keep DEF_HIST_LEN bytes of
+		   history at the beginning of fifo_in. Data from next_in has priority
+		   because it comes later in the data stream. */
+		copy_bytes = from_next_in < DEF_HIST_LEN ? DEF_HIST_LEN-from_next_in : 0;
+		if (copy_bytes > 0)
+			memmove(s->fifo_in, s->fifo_in + s->cur_in - copy_bytes, copy_bytes);
```

Comment:
At the risk of suggesting a premature optimization...
I believe this copy could also make room for `nx_config.cache_threshold` bytes.
This would guarantee that `cache_input()` would always have enough room for itself.

Reply:
I experimented with this, but couldn't get it to pass all tests. I'll give a second shot soon.

Reply:
@tuliom I'm working on an optimization that completely removes these `memmove` calls to reset `fifo_in`. But it still has some bugs, so I'll probably include it in a separate PR.

> I believe this copy could also make room for nx_config.cache_threshold bytes.
Note this copy already makes as much room as possible on `fifo_in`. We only copy the amount of data needed to maintain the history, so at most 32 kB.